jaeger-core/src/main/java/com/uber/jaeger/Tracer.java

```diff
@@ -192,11 +197,32 @@ public void close() {
     @Override
     public io.opentracing.Tracer.SpanBuilder addReference(
         String referenceType, io.opentracing.SpanContext referencedContext) {
-      if (referencedContext instanceof SpanContext
-          && (Utils.equals(referenceType, References.CHILD_OF)
-                      || Utils.equals(referenceType, References .FOLLOWS_FROM))) {
-        this.references.add(new Reference((SpanContext) referencedContext, referenceType));
-        this.baggage.putAll(((SpanContext)referencedContext).baggage());
+
+      if (!(referencedContext instanceof SpanContext)) {
+        return this;
+      }
+
+      // TODO UI currently does not support arbitrary references
+      if (!References.CHILD_OF.equals(referenceType)
+          && !References.FOLLOWS_FROM.equals(referenceType)) {
+        return this;
+      }
+
+      if (parentContext == null && references == null
+          && References.CHILD_OF.equals(referenceType)) {
+        parentContext = (SpanContext) referencedContext;
+      } else {
```

Comment:
This is not sufficient for situations when there is only one parent, so the baggage can be reused.